File: sliding-window-maximum/solution239.go
// https://leetcode.com/problems/sliding-window-maximum/

package main

import (
	"container/list"
	"fmt"
	"math"
)

// T: O(n), S: O(n)
func maxSlidingWindow(nums []int, k int) []int {
	dq := list.New()
	n := len(nums)
	result := make([]int, n-k+1)
	idx := 0
	for i := 0; i < n; i++ {
		for dq.Len() > 0 && nums[i] >= nums[dq.Back().Value.(int)] {
			dq.Remove(dq.Back())
		}
		dq.PushBack(i)
		if i-dq.Front().Value.(int) >= k {
			dq.Remove(dq.Front())
		}
		if i >= k-1 {
			result[idx] = nums[dq.Front().Value.(int)]
			idx++
		}
	}
	return result
}

// T: O(nk), S: O(n)
func maxSlidingWindowBrute(nums []int, k int) []int {
	len := len(nums) - k + 1
	res := make([]int, len)
	p := 0
	for i := 0; i < len; i++ {
		winMax := maxBrute(nums, i, i+k)
		res[p] = winMax
		p++
	}
	return res
}

func maxBrute(nums []int, beg, end int) int {
	max := math.MinInt32
	for i := beg; i < end; i++ {
		if nums[i] > max {
			max = nums[i]
		}
	}
	return max
}

func main() {
	nums1 := []int{1, 3, -1, -3, 5, 3, 6, 7}
	result1 := maxSlidingWindow(nums1, 3)
	fmt.Println(result1) // [3,3,5,5,6,7]
	result2 := maxSlidingWindowBrute(nums1, 3)
	fmt.Println(result2) // [3,3,5,5,6,7]
}

File: sliding-window-maximum/Solution239.java
// https://leetcode.com/problems/sliding-window-maximum/

import java.util.ArrayDeque;
import java.util.Arrays;
import java.util.Deque;

class Solution239 {

    // T: O(n), S: O(n)
    public int[] maxSlidingWindow(int[] nums, int k) {
        Deque<Integer> dq = new ArrayDeque<>();
        int n = nums.length;
        int[] result = new int[n - k + 1];
        int idx = 0;
        for (int i = 0; i < n; i++) {
            while (!dq.isEmpty() && nums[i] >= nums[dq.peekLast()]) {
                dq.pollLast();
            }
            dq.offer(i);
            if (i - dq.peekFirst() >= k) {
                dq.pollFirst();
            }
            if (i >= k - 1) {
                result[idx++] = nums[dq.peekFirst()];
            }
        }
        return result;
    }

    // T: O(nk), S: O(n)
    public int[] maxSlidingWindowBrute(int[] nums, int k) {
        int len = nums.length - k + 1;
        var result = new int[len];
        int p = 0;
        for (int i = 0; i < len; i++) {
            int winMax = maxBrute(nums, i, i + k);
            result[p++] = winMax;
        }
        return result;
    }

    private int maxBrute(int[] nums, int beg, int end) {
        int max = Integer.MIN_VALUE;
        for (int i = beg; i < end; i++) {
            max = Math.max(max, nums[i]);
        }
        return max;
    }

    public static void main(String[] args) {
        var sol = new Solution239();
        var nums1 = new int[] { 1, 3, -1, -3, 5, 3, 6, 7 };
        var result1 = sol.maxSlidingWindow(nums1, 3);
        System.out.println(Arrays.toString(result1)); // [3,3,5,5,6,7]
        var result2 = sol.maxSlidingWindow(nums1, 3);
        System.out.println(Arrays.toString(result2)); // [3,3,5,5,6,7]
    }

}
